### DISCUSSION TITLE
Making a batch from samples with different number of tokens

### DISCUSSION THREAD

**Discussion by Majdoddin** [openai/whisper#753] [Q&A] [answered]
To fine-tune whisper's decoder, I should make a batch of training samples. Each sample has an audio_features tensor, which is the output of encoder, and a tokens tensor. 
Now all audio_features are the same length (30s window), but each sample can have a different number of tokens, and here is the problem: all the tokens tensors should be packed into one batch Tensor, to be passed as argument to model.decoder() (which is mapped to TextDecoder.forward(), which in Turn calls [torch.nn.functional.embedding](https://pytorch.org/docs/stable/generated/torch.nn.functional.embedding.html)). But Tensors of different lengths can't be packed into one tensor :(

@jongwook  How did you do it while training, please?
Is there some "neutral" Token, to pad all tensors to the same size? 
Or should I change the codebase to do the embedding sample by sample?

Similar problem is discussed [here](https://discuss.pytorch.org/t/variable-length-indexing-of-embeddings-without-padding/24708).

**Comment by jongwook** [comment, marked as answer]
You can pad the tensor (to the right) with any token, and the choice does not affect the training because of the autoregressive attention mask used in the decoder. It is also important that the loss is masked accordingly.

In the code, the mask is implemented as the following:

https://github.com/openai/whisper/blob/28769fcfe50755a817ab922a7bc83483159600a9/whisper/model.py#L175-L175
